### DISCUSSION TITLE
`gix-transport` for server

### DISCUSSION THREAD

**Discussion by NicholasLYang** [GitoxideLabs/gitoxide#1299] [Ideas]
### Summary 💡

I'm looking to implement the server portion of the git pack protocol, i.e. stuff like `upload-pack`. As far as I can tell, gitoxide doesn't do this yet. How much of it is implemented? I can only find references to client code in `gix-transport` and `gix-protocol`. 

### Motivation 🔦

I'm looking to write my own git server in Rust, mostly for fun. I'm okay with writing my own git pack serializer/deserializer, but it'd be easier to build off of gitoxide.

**Comment by Byron** [comment]
It would definitely be great to have that!

I recommend to implement `upload-pack` first as it naturally has to deal with the protocol while just using stdin/stdout. There is [a tracking ticket](https://github.com/Byron/gitoxide/issues/734) for that as well.

An initial analysis of the server side was put [into this tracking ticket](https://github.com/Byron/gitoxide/issues/307).

And as it's not mentioned anywhere possibly, there is a `gix free pack create` command which can be used to see how packs can be created in a streaming fashion.

My recommendation would be to use `gitoxide`'s pack creation, and to try to use (and/or adapt) `gix-transport`. The protocol portion in `gix-protocol` might be usable, but the less you try to integrate with these the less you are slowed by their client-centricity. Ideally, you would be able to produce a working `gix upload-pack` by cobbling it together, which would already yield a lot of experience that may one day help to integrate it into the `gix-protocol` and `gix-transport` crates.

In any case, other plumbing crates like `gix-pack` and `gix-revwalk` should already be very usable for use, and probably also the `gix` top-level to more easily interact with a Git repository.

I hope that helps.
